doc/design.md

```diff
@@ -1,51 +1,203 @@
+
+Covers:
+
+* `req~verify-existence-of-files~1`
+
+Needs: impl, utest, itest
+
+### Deleted Files Validator
+
+`dsn~deleted-files-validator~1`
+
+For configuring the deleted files validator we decided for a Java API. For each deleted file one can specify a reason that will be printed as error message. By that one can for example tell that the file was moved.
+
+Covers:
+
+* `req~verify-non-existence-of-files~1`
+
+Needs: impl, utest, itest
+
+### Maven Plugin Validator
+
+`dsn~mvn-plugin-validator~1`
+
+We decided to implement the verification of the pom plugins using a combination of xml files and Java code. The XML files (in resources/mvn_templates) contain the XML of a typical plugin definition. The Java class defines which XML properties must be like define in the template.
+
+This approach allows us to quickly add new templates by copying an exemplary plugin definition from a pom.xml file, and the defining which parts are mandatory using Java.
+
+Covers:
+
+* `req~verify-mvn-plugins~1`
+
+Needs: impl, utest, itest
+
+### Mvn Dependency Validator
+
+`dsn~mvn-dependency-validator~1`
+
+We implemented the validation for dependency declaration using Java code. That allows us to add specific validations for certain dependencies. For example for the jacoco dependency we need to validate the scope.
+
+However, implementing it purely via configuration would probably also be a good solution. In the future maybe even a better one, depending on how many dependencies should get validated.
+
+Covers:
+
+* `req~verify-maven-dependencies~1`
+
+Needs: impl, utest, itest
+
+### Dependency Changes Section in changes_x.x.x.md File Validator
+
+`dsn~dependency-section-in-changes_x.x.x.md-file-validator~1`
+
+The basic approach for this validator is to generate the markdown for the `Dependency Changes` section and then compare the content of the project's file with that generated content. In case they differ the validation fails. The automatic fix of the validation error is to override the section with the generated one.
+
+For this feature we need to generate the list of dependency changes since the last release. We decided to generate this list by comparing the dependencies of the last release with the dependencies of this release.
+
+For both versions we [read the project's dependencies](#reading-project-dependencies) and create a java model that we then compare to extract the diffs.
+
+For getting the last release we use git. We detect the previous release by walking through the commits on the current branch in reverse order. If we find a tag that looks like a version number we extract the pom.xml file from this commit and analyze its dependencies.
+
+We separated the crawling of the dependencies from the rendering of the table, so that we can add different crawlers for other languages or build systems in the future.
```

Comment:
```suggestion
We separated the crawling of the dependencies from the rendering of the table so that we can add different crawlers for other languages or build systems in the future.
```